File: Week12/LeetCode-0210-1.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        // Create DAG from prerequisites
        vector<vector<int>> G(numCourses);
        vector<int> inDegree(numCourses, 0);
        
        // Edge: prerequisites[i][1]-->prerequisites[i][0]
        for (int i=0; i<prerequisites.size(); i++){
            G[prerequisites[i][1]].push_back(prerequisites[i][0]);
            inDegree[prerequisites[i][0]]++;
        } // end-for
        
        queue<int> Q; // Queue of nodes with in-degree == 0
        for (int i=0; i<numCourses; i++)
            if (inDegree[i] == 0)
                Q.push(i);

        vector<int> result;
        while (!Q.empty()){
            auto u = Q.front(); Q.pop();
            result.push_back(u);
            
            // Reduce In-Degree of all vertices adjacent to u by 1
            for (auto v: G[u]){
                if (--inDegree[v] == 0)
                    Q.push(v);
            } // end-for
        } // end-while
        
        return result.size() == numCourses? result: vector<int>{};
    } // end-findOrder
};

File: Week12/LeetCode-0210-1.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # Create DAG from prerequisites
        G = []
        for i in range(numCourses): G.append([])
        inDegree = [0]*numCourses
        
        # Edge: prerequisites[i][1]-->prerequisites[i][0]
        for i in range(len(prerequisites)):
            G[prerequisites[i][1]].append(prerequisites[i][0])
            inDegree[prerequisites[i][0]] += 1

        Q = []  # Queue of nodes with in-degree == 0
        for i in range(numCourses):
            if inDegree[i] == 0:
                Q.append(i)

        result = []
        while len(Q) > 0:
            u = Q.pop(0)
            result.append(u)
            
            # Reduce In-Degree of all vertices adjacent to u by 1
            for v in G[u]:
                inDegree[v] -= 1
                if inDegree[v] == 0:
                    Q.append(v)        
        return result if len(result) == numCourses else []
                